src/Covenant/ASG.hs

```diff
@@ -677,6 +700,139 @@ app fId argRefs instTys = do
+          )
+        . sequence
+  {- The procedures for handling a typed declared as Opaque and a "normal" type are totally different.
+
+     For Opaque types, we just have to check that the "ConstructorName" we get corresponds to a constructor of
+     PlutusData, then validate that the arguments conform with that PlutusData constructor.
+  -}
+  case thisTyDecl of
+    OpaqueData _ opaqueCtorSet -> do
+      checkOpaqueArgs opaqueCtorSet renamedFieldTypes
+      refTo $ AValNode (Datatype tyName mempty) (DataConstructorInternal tyName ctorName fields)
+    DataDeclaration _ count ctors _ -> do
+      -- First we check that the arity of the constructor is equal to the number of fields in the decl.
+      checkFieldArity (Vector.length fields) thisTyInfo
+      -- Then we resolve the supplied field Refs, rename, and throw an error if we're passed an error node.
+
+      -- Then we construct the return type.
+      resultThunk <- mkResultThunk count ctors renamedFieldTypes
+      -- Then we undo the renaming.
+      restored <- undoRenameM resultThunk
+      -- Then we check the compatibility of the arguments with the datatype's encoding.
+      asks (view #datatypeInfo) >>= \dti -> checkEncodingWithInfo dti restored
+      -- Finally, if nothing has thrown an error, we return a reference to our result node, decorated with the
+      -- return type we constructed.
+      refTo $ AValNode restored (DataConstructorInternal tyName ctorName fields)
+  where
+    {- Constructs the result type of the introduction form. Arguments are:
+         1. The count (number of tyvars) from the data declaration.
+         2. The vector of constructors from the data declaration.
+         3. The (renamed and resolved) vector of supplied arguments.
+
+       The procedure goes like:
+         1. Extract the argument types from the constructor in the declaration. Any tyvars here
+            *have* to be Unifiable (unless something has slipped past the kind checker) -
+            data declarations have atomic, independent scopes.
+         2. Unify those with the actual, supplied field types, yielding a set of substitutions.
+         3. Construct a fully abstract (i.e. parameterized only by unifiable type variables) representation of the
+            type constructor.
+         4. Apply those substitutions to the abstract type constructor.
+         5. Wrap the "concretified" type constructor in a thunk and use `fixUp` to sort out the `Count` and
+            indices.
+    -}
+    mkResultThunk :: Count "tyvar" -> Vector (Constructor Renamed) -> Vector (ValT Renamed) -> m (ValT Renamed)
+    mkResultThunk count' declCtors fieldArgs' = do
+      declCtorFields <- Vector.toList . view #constructorArgs <$> findConstructor declCtors
+      subs <- unifyFields declCtorFields fieldArgs
+      let tyConAbstractArgs
+            | count == 0 = []
+            | otherwise = Abstraction . Unifiable . fromJust . preview intIndex <$> [0 .. (count - 1)]
```

Comment:
You can avoid branching here by using `[0, 1 .. count - 1]`, as this will be empty if `count` is `0`. You can also avoid a `fromJust` here by using `mapMaybe` instead of `<$>`.